app/src/main/java/dev/msfjarvis/lobsters/data/remote/LobstersPagingSource.kt

```diff
@@ -2,17 +2,30 @@ package dev.msfjarvis.lobsters.data.remote
 
 import androidx.paging.PagingSource
 import dev.msfjarvis.lobsters.data.api.LobstersApi
-import dev.msfjarvis.lobsters.model.LobstersPost
+import dev.msfjarvis.lobsters.data.local.LobstersPost
+import dev.msfjarvis.lobsters.data.repo.LobstersRepository
 import javax.inject.Inject
 
 class LobstersPagingSource @Inject constructor(
   private val lobstersApi: LobstersApi,
+  private val lobstersRepository: LobstersRepository,
 ) : PagingSource<Int, LobstersPost>() {
 
   override suspend fun load(params: LoadParams<Int>): LoadResult<Int, LobstersPost> {
     return try {
       val page = params.key ?: 1
-      val posts = lobstersApi.getHottestPosts(page)
+      val posts = lobstersApi.getHottestPosts(page).map { post ->
+        // We can replace these two lines below with a getOrInsertPost repository method.
```

Comment:
Why don't we then?

Reply:
Because it is still WIP :eyes:

Reply:
Also, one problem that I've noticed is that we don't have a single source of truth for the data. Some data is coming from the api and other is coming from the DB.

I think we should make the DB the single source of truth and load all data from there.

Reply:
I'm not too sure about having the main stream being persisted to the database. It's nice as an offline fallback but invalidation is tricky.

Reply:
We are fetching the list directly in `PagingSource` and there's no underlying list source that we can update and then call `invalidate` on. If we don't go the DB route, we'll need something to back the list and the changes to that backing property will then invalidate the `PagingSource` and update the UI.

Reply:
Paging already handles updating the UI though? That's its entire point.

Reply:
AFAIK there's two cases in which Paging would update the UI
 
- When we reach the end of the page
- When there's a change in underlying data source

In the second case, we don't really have a source to update once we the `is_saved` property is updated and hence the `PagedList` never actually changes.

Reply:
Hm yeah going through the docs it does seem to require a concrete local source as well. Let's add a database for posts but clear it on ViewModel creation so it's empty when paging starts loading.